File: test/sort.go
// sort.go
// Generate N random strings (length 33), store in []string, sort and measure time.
// Usage: go run sort.go N [runs] [seed]

package main

import (
	crand "crypto/rand"
	"encoding/binary"
	"flag"
	"fmt"
	"math/rand"
	"os"
	"sort"
	"time"
)

func genRandomString(lenBytes int, r *rand.Rand) string {
	const charset = "0123456789abcdef"
	b := make([]byte, lenBytes)
	for i := 0; i < lenBytes; i++ {
		b[i] = charset[r.Intn(len(charset))]
	}
	return string(b)
}

func main() {
	flag.Parse()
	args := flag.Args()
	if len(args) < 1 {
		fmt.Fprintf(os.Stderr, "Usage: %s N [runs] [seed]\n", os.Args[0])
		os.Exit(1)
	}

	var N int
	_, err := fmt.Sscan(args[0], &N)
	if err != nil || N < 0 {
		fmt.Fprintf(os.Stderr, "Invalid N: %s\n", args[0])
		os.Exit(1)
	}

	runs := 1
	if len(args) >= 2 {
		_, err := fmt.Sscan(args[1], &runs)
		if err != nil || runs < 1 {
			runs = 1
		}
	}

	var seed int64
	if len(args) >= 3 {
		_, err := fmt.Sscan(args[2], &seed)
		if err != nil {
			seed = time.Now().UnixNano()
		}
	} else {
		// generate a random seed from crypto/rand for better variance
		var b [8]byte
		if _, err := crand.Read(b[:]); err == nil {
			seed = int64(binary.LittleEndian.Uint64(b[:]))
		} else {
			seed = time.Now().UnixNano()
		}
	}

	fmt.Printf("N=%d runs=%d seed=%d\n", N, runs, seed)

	const KEY_LEN = 33
	const VAL_LEN = 600

	type Record struct {
		Key   string
		Value string
	}

	for runc := 0; runc < runs; runc++ {
		rng := rand.New(rand.NewSource(seed + int64(runc)))
		arr := make([]Record, 0, N)
		for i := 0; i < N; i++ {
			rec := Record{
				Key:   genRandomString(KEY_LEN, rng),
				Value: genRandomString(VAL_LEN, rng),
			}
			arr = append(arr, rec)
		}

		t0 := time.Now()
		sort.Slice(arr, func(i, j int) bool { return arr[i].Key < arr[j].Key })
		t1 := time.Now()
		dur := t1.Sub(t0).Seconds() * 1000.0
		fmt.Printf("run %d: sort time = %.6f ms\n", runc+1, dur)
		if len(arr) > 0 {
			fmt.Printf("first_key=%s last_key=%s\n", arr[0].Key[:6], arr[len(arr)-1].Key[:6])
		}
	}
}

File: test/sort.cpp
// sort.cpp
// Generate N random records (key length 33, value length 600), store in vector<Record>, sort by key and measure time.
// Usage: sort_test N [runs] [seed]

#include <algorithm>
#include <chrono>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using namespace std;

struct Record {
	string key;   // 33 bytes
	string value; // 600 bytes
};

static string gen_random_hex(size_t len, std::mt19937_64 &rng) {
	static const char charset[] = "0123456789abcdef"; // 16 chars
	std::uniform_int_distribution<int> dist(0, 15);
	string s;
	s.resize(len);
	for (size_t i = 0; i < len; ++i) s[i] = charset[dist(rng)];
	return s;
}

int main(int argc, char **argv) {
	if (argc < 2) {
		cerr << "Usage: " << argv[0] << " N [runs] [seed]\n";
		return 1;
	}

	size_t N = 0;
	try { N = stoull(argv[1]); } catch (...) { cerr << "Invalid N\n"; return 1; }

	int runs = 1;
	if (argc >= 3) { try { runs = stoi(argv[2]); } catch (...) { runs = 1; } if (runs < 1) runs = 1; }

	unsigned long long seed = (unsigned long long)chrono::high_resolution_clock::now().time_since_epoch().count();
	if (argc >= 4) { try { seed = stoull(argv[3]); } catch (...) { } }

	cout << "N=" << N << " runs=" << runs << " seed=" << seed << "\n";

	const size_t KEY_LEN = 33;
	const size_t VAL_LEN = 600;

	vector<double> times;
	times.reserve(runs);

	for (int r = 0; r < runs; ++r) {
		std::mt19937_64 rng(seed + (unsigned long long)r);

		vector<Record> v;
		v.reserve(N);

		for (size_t i = 0; i < N; ++i) {
			Record rec;
			rec.key = gen_random_hex(KEY_LEN, rng);
			rec.value = gen_random_hex(VAL_LEN, rng);
			v.emplace_back(std::move(rec));
		}

		auto t0 = chrono::steady_clock::now();
		sort(v.begin(), v.end(), [](const Record &a, const Record &b) { return a.key < b.key; });
		auto t1 = chrono::steady_clock::now();

		double ms = chrono::duration_cast<chrono::duration<double, std::milli>>(t1 - t0).count();
		times.push_back(ms);

		cout << "run " << (r + 1) << ": sort time = " << ms << " ms\n";
		if (!v.empty()) {
			cout << "first_key=" << v.front().key.substr(0,6) << " last_key=" << v.back().key.substr(0,6) << "\n";
		}
	}

	double sum = 0.0;
	for (double t : times) sum += t;
	double avg = sum / times.size();
	cout << "average sort time = " << avg << " ms\n";

	return 0;
}

